File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.py
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        import copy        
        #backtracking. Can keep track of only one subset and can infer the rest

        def dfs(i, subs1):

            #end case. Check the end conditions here
            if i == len(nums):
                s1, s2 = 0, 0
                
                if not subs1 or len(subs1) == 0:
                    return False

                print("Subs1 <>")
                for i in range(len(nums)):
                    
                    if i in subs1:
                        s1 += nums[i]
                        print("\tSubs1 elements are: ", nums[i])
                    else:
                        s2 += nums[i]
                    
                if s1 == s2:
                    return True
                return False
            
            #not take
            t1 = dfs(i+1, subs1)

            #take
            temp = copy.deepcopy(subs1)
            temp.append(i)
            t2 = dfs(i+1, temp)
            
            return t1 or t2

        return dfs(0, [])

File: Data Structures & Algorithms/largest-divisible-subset/submission-1.cpp
class Solution {
public:
    
    int max_len = 0;
    vector<int> res = {};
    
    void _dfs(vector<int> &nums, int n, vector<int>& subset, int i) {
        if (i >= n) {
            // check if this subset is higher or not
            if (max_len < subset.size()) {
                res = subset;
                max_len = subset.size();
            }
            return;
        }

        // take or not
        _dfs(nums, n, subset, i+1);

    
        if (subset.empty() || nums[i] % subset.back() == 0) {
            // take this
            subset.push_back(nums[i]);
            _dfs(nums, n, subset, i+1);
            subset.pop_back();
        }

        return;
    }
    
    vector<int> largestDivisibleSubset(vector<int>& nums) {
        sort(nums.begin(), nums.end()); vector<vector<int>> cache = {}; int n = nums.size(); vector<int> temp = {};
        _dfs(nums, n, temp, 0);
        return res;
    }
};